### PR TITLE
[flang][OpenMP] Add support for copyprivate

### PR DESCRIPTION
**Pull request by luporl**
Add initial handling of OpenMP copyprivate clause in Flang.

When lowering copyprivate, Flang generates the copy function
needed by each variable and builds the appropriate
omp.single's CopyPrivateVarList.

This is patch 3 of 4, to add support for COPYPRIVATE in Flang.
Original PR: https://github.com/llvm/llvm-project/pull/73128

### CODE REVIEW

File: flang/lib/Lower/Bridge.cpp [hunk 1]
```diff
@@ -1092,6 +1040,79 @@ class FirConverter : public Fortran::lower::AbstractConverter {
...
+                                                stmtCtx);
+      stmtCtx.finalizeAndReset();
+    } else if (lhs.getBoxOf<fir::CharBoxValue>()) {
+      fir::factory::CharacterExprHelper{*builder, loc}.createAssign(lhs, rhs);
+    } else {
+      auto loadVal = builder->create<fir::LoadOp>(loc, fir::getBase(rhs));
+      builder->create<fir::StoreOp>(loc, loadVal, fir::getBase(lhs));
+    }
+  }
+
```
Comment (clementval) [inline flang/lib/Lower/Bridge.cpp:1115]: Could this be in OpenMP.cpp? It would remove the need to add a function to the converter.

**Comment by luporl** [comment]
ping

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

File: flang/lib/Lower/OpenMP.cpp [hunk 2]
```diff
@@ -1178,6 +1183,102 @@ class ReductionProcessor {
...
+  } else if (auto bty = mlir::dyn_cast<fir::BoxType>(ty)) {
+    inBox = true;
+    typeScan(bty.getEleTy());
+  } else if (auto cty = mlir::dyn_cast<fir::CharacterType>(ty)) {
+    charLen = cty.getLen();
+  } else if (auto hty = mlir::dyn_cast<fir::HeapType>(ty)) {
+    typeScan(hty.getEleTy());
+  } else if (auto pty = mlir::dyn_cast<fir::PointerType>(ty)) {
+    typeScan(pty.getEleTy());
+  }
```
[the remark this answers is not in the document]
Comment (luporl) [reply flang/lib/Lower/OpenMP.cpp:1228]: Done.

**Pull request merged**